packages/xo-server-sdn-controller/src/index.js

```diff
@@ -717,6 +676,68 @@ class SDNController extends EventEmitter {
     }
   }
 
+  async _addHostToPoolNetworks(host) {
+    const xapi = host.$xapi
+
+    const tunnels = filter(xapi.objects.all, { $type: 'tunnel' })
+    for (const tunnel of tunnels) {
+      const accessPIF = await xapi._getOrWaitObject(tunnel.access_PIF)
+      if (accessPIF.host !== host.$ref) {
+        continue
+      }
+
+      const poolNetwork = find(this._poolNetworks, {
+        network: accessPIF.network,
+      })
+      if (poolNetwork == null) {
+        continue
+      }
+
+      if (accessPIF.currently_attached) {
+        continue
+      }
+
+      log.debug(
+        `Pluging PIF: '${accessPIF.device}' for host: '${
+          host.name_label
+        }' on network: '${accessPIF.$network.name_label}'`
+      )
+      try {
+        await xapi.call('PIF.plug', accessPIF.$ref)
+      } catch (error) {
+        log.error(
+          `XAPI error while pluging PIF: '${accessPIF.device}' on host: '${
+            host.name_label
+          }' for network: '${accessPIF.$network.name_label}'`
+        )
+      }
+
+      const starCenter = await xapi._getOrWaitObject(poolNetwork.starCenter)
+      await this._addHostToNetwork(host, accessPIF.$network, starCenter)
+    }
+  }
+
+  async _hostUnreachable(host) {
+    const poolNetworks = filter(this._poolNetworks, { starCenter: host.$ref })
+    for (const poolNetwork of poolNetworks) {
+      const network = await host.$xapi._getOrWaitObject(poolNetwork.network)
+      log.debug(
+        `Star center host: '${host.name_label}' of network: '${
+          network.name_label
+        }' in pool: '${
+          host.$pool.name_label
+        }' is no longer reachable, electing a new host`
+      )
+
+      const newCenter = await this._electNewCenter(network, true)
+      poolNetwork.starCenter = newCenter != null ? newCenter.$ref : null
```

Comment:
Can `newCenter` be `null` and `undefined`?

IMHO it's better to stick with one.

Reply:
Only null normally. I should use `!== null` then?

Reply:
IMHO it's better to stick with `undefined` because it's the only value that can happen by itself (contrary to `null` which is an explicit value).

Also if you install [`@babel/plugin-proposal-optional-chaining`](https://babeljs.io/docs/en/babel-plugin-proposal-optional-chaining), you will be able to use the optional chaining syntax: `newCenter?.$ref`.

Reply:
Love the `foo?.bar` syntax!